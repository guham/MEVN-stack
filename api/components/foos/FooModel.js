const mongoose = require('mongoose');

const schema = new mongoose.Schema(
  {
    name: {
      type: String,
      required: true,
      unique: true,
    },
  },
  {
    timestamps: true,
  },
);

module.exports = mongoose.model('Foo', schema);
