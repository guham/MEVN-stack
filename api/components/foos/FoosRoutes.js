const { Router } = require('express');
const { makeInvoker } = require('awilix-express');
const { asyncMiddleware } = require('../../middlewares/errorHandlers');
const foosController = require('./FoosController');

const router = Router();
const api = makeInvoker(foosController);

router.get('/', asyncMiddleware(api('index')));
router.post('/add', asyncMiddleware(api('add')));
router.get('/test', asyncMiddleware(api('test')));
router.get('/:id', asyncMiddleware(api('get')));

module.exports = router;
