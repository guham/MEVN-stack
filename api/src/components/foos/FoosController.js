const foosController = ({ foosService }) => ({
  index: async (req, res, next) => {
    const foos = await foosService.getAll();
    res.json(foos);
  },

  add: async (req, res, next) => {
    const { name } = req.body;
    const foo = await foosService.add([
      { name },
    ]);
    res.status(201).json(foo);
  },

  get: async (req, res, next) => {
    const { id } = req.params;
    const foo = await foosService.getById(id);
    res.json(foo);
  },

  test: async (req, res, next) => {
    res.json({ data: 'This is a message fetched from the API' });
  },
});

module.exports = foosController;
