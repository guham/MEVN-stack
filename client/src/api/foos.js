import { createClient } from './client';

const client = createClient();

export default {
  /**
   * @returns {Array<Object>} array of foos
   */
  async fetchFoos() {
    const response = await client.get('/api/foo');
    return response.data;
  },
  /**
   * @param {string} name
   * @returns {Object} added foo
   */
  async addFoo(name) {
    const response = await client.post('/api/foo/add', { name });
    return response.data;
  },
};
