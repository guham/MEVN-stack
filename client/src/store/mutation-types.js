export const ADD_FOO = 'ADD_FOO';
export const UPDATE_FOO_NAME = 'UPDATE_FOO_NAME';
export const FETCH_FOOS = 'FETCH_FOOS';

export const SET_ERROR = 'SET_ERROR';
export const RESET_ERROR = 'RESET_ERROR';

export const SET_USER_AUTHENTICATED = 'SET_USER_AUTHENTICATED';
export const SET_USER_UNAUTHENTICATED = 'SET_USER_UNAUTHENTICATED';
export const SET_USER_IS_SIGNING_IN = 'SET_USER_IS_SIGNING_IN';
export const SET_ACCESS_TOKEN = 'SET_ACCESS_TOKEN';
export const SET_REFRESH_TOKEN = 'SET_REFRESH_TOKEN';
export const SET_EXPIRATION_DATE = 'SET_EXPIRATION_DATE';

export const ADD_NOTIFICATION = 'ADD_NOTIFICATION';
export const REMOVE_NOTIFICATION = 'REMOVE_NOTIFICATION';
