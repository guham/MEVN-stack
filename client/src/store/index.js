import Vue from 'vue';
import Vuex from 'vuex';
import foos from './modules/foos';
import user from './modules/user';

Vue.use(Vuex);

const store = new Vuex.Store({
  strict: process.env.NODE_ENV !== 'production',
  modules: {
    foos,
    user,
  },
});

/* eslint-disable global-require */
if (module.hot) {
  module.hot.accept([
    './modules/foos',
    './modules/user'], () => {
    const newFoos = require('./modules/foos').default;
    const newUser = require('./modules/user').default;
    store.hotUpdate({
      modules: {
        newFoos,
        newUser,
      },
    });
  });
}

export default store;
