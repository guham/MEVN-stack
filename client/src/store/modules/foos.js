import _get from 'lodash/get';
import foosClient from '@/api/foos';
import * as types from '../mutation-types';

const state = {
  foos: [],
  name: '',
  error: {},
};

const getters = {
  count: state => state.foos.length,
  isValidName: state => state.name.length > 0,
  hasError: state => Object.keys(state.error).length > 0,
  errorMessage: (state, getters) => (getters.hasError ? _get(state.error, 'data.message', _get(state.error, 'message', '')) : ''),
};

const mutations = {
  [types.FETCH_FOOS](state, foos) {
    state.foos = foos;
  },

  [types.ADD_FOO](state, foo) {
    state.foos.push(foo);
  },

  [types.UPDATE_FOO_NAME](state, name) {
    state.name = name.trim();
  },

  [types.SET_ERROR](state, error) {
    state.error = error;
  },

  [types.RESET_ERROR](state) {
    state.error = {};
  },
};

const actions = {
  async fetchFoos({ commit }) {
    try {
      const foos = await foosClient.fetchFoos();
      commit(types.FETCH_FOOS, foos);
    } catch (error) {
      commit(types.SET_ERROR, error);
    }
  },

  async addFoo({ commit, state, getters }) {
    if (!getters.isValidName) {
      return;
    }
    try {
      const foo = await foosClient.addFoo(state.name);
      commit(types.ADD_FOO, foo);
      commit(types.UPDATE_FOO_NAME, '');
      commit(types.RESET_ERROR);
    } catch (error) {
      commit(types.SET_ERROR, error);
    }
  },
};

export default {
  namespaced: true,
  state,
  getters,
  mutations,
  actions,
};
