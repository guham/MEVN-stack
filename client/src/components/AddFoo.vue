<template>
  <div class="flex justify-center pt-6">
    <form
      class="w-full max-w-xs"
      @submit.prevent>
      <div class="md:flex md:items-center mb-6">
        <div class="md:w-1/3">
          <label
            for="add-foo"
            class="block text-grey font-bold md:text-right mb-1 md:mb-0 pr-4">Foo</label>
        </div>
        <div class="md:w-2/3">
          <input
            id="add-foo"
            v-model="name"
            :placeholder="placeholder"
            class="input"
            type="text"
            @keyup.enter="addFoo()">
        </div>
      </div>
      <div
        v-if="hasError"
        id="error"
        class="md:flex md:items-center mb-6">
        <p class="text-red text-sm text-justify italic">{{ errorMessage }}</p>
      </div>
      <div class="md:flex md:items-center">
        <div class="md:w-1/3"/>
        <div class="md:w-2/3">
          <button
            class="btn btn-green px-4"
            type="button"
            @click="addFoo()">{{ $t('add') }}</button>
        </div>
      </div>
    </form>
  </div>
</template>

<i18n>
{
  "en": {
    "add": "Add",
    "default-placeholder": "Default placeholder"
  },
  "fr": {
    "add": "Ajouter",
    "default-placeholder": "Placeholder par d√©faut"
  },
  "es": {
    "add": "Agregar",
    "default-placeholder": "Placeholder por defecto"
  }
}
</i18n>

<script>
import { createNamespacedHelpers } from 'vuex';
import * as types from '@/store/mutation-types';

const { mapMutations, mapActions, mapGetters } = createNamespacedHelpers('foos');

export default {
  name: 'AddFoo',
  props: {
    placeholder: {
      type: String,
      default() {
        return this.$t('default-placeholder');
      },
    },
  },

  computed: {
    name: {
      get() {
        return this.$store.state.foos.name;
      },
      set(name) {
        this[types.UPDATE_FOO_NAME](name);
      },
    },
    ...mapGetters([
      'hasError',
      'errorMessage',
    ]),
  },

  methods: {
    ...mapMutations([
      types.UPDATE_FOO_NAME,
    ]),
    ...mapActions([
      'addFoo',
    ]),
  },
};
</script>
