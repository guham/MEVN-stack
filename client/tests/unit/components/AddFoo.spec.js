import Vuex from 'vuex';
import AddFoo from '@/components/AddFoo.vue';
import foos from '@/store/modules/foos';
import * as types from '@/store/mutation-types';
import factory from '../factory';

beforeEach(() => {
  jest.resetModules();
  jest.clearAllMocks();
});

describe('AddFoo.vue', () => {
  let store;
  let actions;
  let mutations;

  beforeEach(() => {
    actions = {
      addFoo: jest.fn(),
    };

    mutations = {
      [types.UPDATE_FOO_NAME]: jest.fn(),
    };

    store = new Vuex.Store({
      modules: {
        foos: {
          namespaced: foos.namespaced,
          state: foos.state,
          getters: foos.getters,
          mutations,
          actions,
        },
      },
    });
  });

  test('renders default props.placeholder', () => {
    const wrapper = factory(AddFoo, {}, {}, [], store);
    expect(wrapper.find('input').element.placeholder).toBe('Default placeholder');
  });

  test('renders props.placeholder when passed', () => {
    const placeholder = 'custom placeholder';
    const wrapper = factory(AddFoo, { placeholder }, {}, [], store);
    expect(wrapper.find('input').element.placeholder).toBe(placeholder);
  });

  test('renders state.name default value in input', () => {
    const wrapper = factory(AddFoo, {}, {}, [], store);
    expect(wrapper.find('input').element.value).toBe(foos.state.name);
  });

  test('should not display error panel when there are no errors', () => {
    const wrapper = factory(AddFoo, {}, {}, [], store);
    expect(wrapper.vm.hasError).toBeFalsy();
    expect(wrapper.find('#error').length).toBeFalsy();
  });

  test('should display error panel when there is an error', () => {
    store = new Vuex.Store({
      modules: {
        foos: {
          namespaced: foos.namespaced,
          state: foos.state,
          getters: {
            error: () => ({
              data: {
                message: 'error message',
              },
            }),
            hasError: () => true,
          },
          mutations,
          actions,
        },
      },
    });

    const wrapper = factory(AddFoo, {}, {}, [], store);
    expect(wrapper.vm.hasError).toBeTruthy();
    expect(wrapper.find('#error').isVisible()).toBeTruthy();
    expect(wrapper.find('#error > p').text()).toBe('error message');
  });

  test('calls store mutation `UPDATE_FOO_NAME` when input value is updated', () => {
    const value = 'new value';
    const wrapper = factory(AddFoo, {}, {}, [], store);
    const input = wrapper.find('input');
    input.element.value = value;
    input.trigger('input');
    expect(mutations.UPDATE_FOO_NAME).toHaveBeenCalled();
  });

  test('calls store action `addFoo` when button is clicked', () => {
    const wrapper = factory(AddFoo, {}, {}, [], store);
    wrapper.find('button').trigger('click');
    expect(actions.addFoo).toHaveBeenCalled();
  });

  test('calls store action `addFoo` when event `keyup.enter` is fired', () => {
    const wrapper = factory(AddFoo, {}, {}, [], store);
    wrapper.find('input').trigger('keyup.enter');
    expect(actions.addFoo).toHaveBeenCalled();
  });

  test('has the expected html structure', () => {
    const wrapper = factory(AddFoo, {}, {}, [], store);
    const template = wrapper.html();
    expect(template).toMatchSnapshot();
  });
});
