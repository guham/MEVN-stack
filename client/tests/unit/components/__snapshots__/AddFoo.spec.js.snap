// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddFoo.vue has the expected html structure 1`] = `
<div class="flex justify-center pt-6">
  <form class="w-full max-w-xs">
    <div class="md:flex md:items-center mb-6">
      <div class="md:w-1/3"><label for="add-foo" class="block text-grey font-bold md:text-right mb-1 md:mb-0 pr-4">Foo</label></div>
      <div class="md:w-2/3"><input id="add-foo" placeholder="default-placeholder" type="text" class="input"></div>
    </div>
    <!---->
    <div class="md:flex md:items-center">
      <div class="md:w-1/3"></div>
      <div class="md:w-2/3"><button type="button" class="btn btn-green px-4">add</button></div>
    </div>
  </form>
</div>
`;
